config/Migrations/20171013133145_Utf8mb4Fix.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+use Cake\Error\Debugger;
+use Phinx\Db\Adapter\MysqlAdapter;
+use Phinx\Migration\AbstractMigration;
+
+class Utf8mb4Fix extends AbstractMigration {
+
+	/**
+	 * Change Method.
+	 *
+	 * Write your reversible migrations using this method.
+	 *
+	 * More information on writing migrations is available here:
+	 * http://docs.phinx.org/en/latest/migrations.html#the-abstractmigration-class
+	 *
+	 * @return void
+	 */
+	public function change() {
+		$table = $this->table('queue_processes');
+		$table->changeColumn('pid', 'text', [
+			'encoding' => 'ascii',
+			'collation' => 'ascii_bin',
+		]);
+
+		$table = $this->table('queued_jobs');
+		$table->changeColumn('job_type', 'text', [
+			'encoding' => 'utf8mb4',
+			'collation' => 'utf8mb4_bin',
+		]);
+		$table->changeColumn('data', 'text', [
+			'encoding' => 'utf8mb4',
+			'collation' => 'utf8mb4_bin',
+		]);
+		$table->changeColumn('job_group', 'text', [
+			'encoding' => 'utf8mb4',
+			'collation' => 'utf8mb4_bin',
```

Comment:
I usually use the linked utf8mb4_unicode_ci for normal text.
What is the upside downside of the _bin one?

Reply:
IMHO one should always use bin over ci.
- bin = default
- ci = case insensitive (where case can clearly be derived... utf8 has a lot of characters where I would not know if there is really "casing"... or just "2-casing")

AFAIK case insensitive does not improve storage (as it does indeed store full utf8mb4) but it may be faster for look-ups.

In any case I would always use bin over ci unless for reasons of performance optimization.
In a way MySQL's default charset/collation is broken in 3 ways:
- latin instead of utf8mb4
- sweden instead of non-specific ;-)
- ci instead of bin
:trollface:

Reply:
Hehe, I much prefer ci over bin (if bin is case sensitive) for normal text as here human interpretation means FOO === foo from the input/output perspective.
Having it allowing different casings here often makes other problems appear.
For pure data fields it is probably fine to switch to cs.

Reply:
Well I can change it no problem because it matters little here.
Personally I am convinced that case sensitivity is a good thing in databases. In file systems... well maybe, maybe not :P.
The reasons are:
- typing A and a is two different things
- comparing A to a should be two different things

p.s.: ci is a weak type and that it is so common is a relic of the mysql era where being fast mattered most (MYISAM = no transactions, no foreign key integrity; case-insensitive indexes so that web searches are as fast as possible, utf8 3-byte instead of the default 4-byte because it takes less storage, etc. etc)

Changing...

Reply:
I think the exact opposite: Human input by default is not CS, always CI. You only need to treat FOO and foo differently when you are in the computer interpreted area.
Thus postgres CS by default is often very dangerous and leads to a lot of unique key violations just because the field was set wrong by default.
Best example: email. Now people can register with 10x the same email and different casings and the system is all good here. Typical use case of the wrong defaults from the start.
So ALWAYS CI unless there is a reason for CS for human input IMO.
And DB columns almost always have human input, especially the varchar ones.

So I would here keep the better strictness (CI) vs. the less strictness (CS) to avoid exactly the issues that you otherwise get!

Reply:
> I think the exact opposite: Human input by default is not CS, always CI. 
(Western) written human input is always case-sensitive.
Spoken input is case insensitive, indeed, for comparison.

Looking up data should be as strict as possible by default. You can then, by design, relax the strictness.

Say convert human audio input with AI to text and then run it against case-insensitive data.

If our input was case insensitive then we'd write all our letters lower-cased. Anyway this point is moot (here, cakephp docs for instance should always recommend utf8mb4 and binary unless you only need case-insensitive or the 3-bytes subset that works for many things). I prefer strict typing, using ci columns makes things weak by default.
It is not required for this plugin however (just the sane, good default, like always using `mb_*` unless you really want to compare bytes instead of characters); so I have changed it already and it. Anything else left to do?

Right now if people rely on the data in this plugin's table they could have malformed data already and I hope it improves things and makes them reliable.

Reply:
A good example of using bad-practises by default is:
- Open up Google Chrome 61
- Visit http://www.utf8-zeichentabelle.de/unicode-utf8-table.pl?number=1024&htmlent=1
- Hit CTRL+F (or CMD+F)
- Enter "ö" - it will find o, O, ö, Ö, Ø and whatnot, there is not even a way to disable the stupidity. I can't find what I am looking for (but then Chrome is only my dev-box browser, for a reason)